torch/_refs/__init__.py

```diff
@@ -4728,21 +4728,23 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
     if isinstance(value, Number):
         value_type = type(value)
     else:
-        # NOTE: Could not use value = item(value) as it resulted in
-        # RuntimeError: Cannot cast FakeTensor(cpu) to number
         value_ndim = value.ndim
         check(
             value_ndim == 0,
             lambda: f"only supports a 0-dimensional value tensor, but got tensor with {value_ndim} dimension",
         )
         # `masked_fill` allows cpu scalar to be moved to cuda but not otherwise.
+        if a.device.type == "cuda" and value.device.type == "cpu":
+            # NOTE: Could not use value = item(value) as it resulted in
+            # RuntimeError: Cannot cast FakeTensor(cpu) to number
+            value = prims.device_put(value, a.device)
```

Comment:
I think this is no longer needed, as `where` accepts 0d cpu tensors

Reply:
I tried this and got 
> FAILED test/inductor/test_torchinductor.py::CudaTests::test_masked_fill_promotion_cuda - RuntimeError: CUDA error: an illegal memory access was encountered

Reply:
Ok looks like `where` inductor lowering lacks those device_put calls, we should fix that, but `masked_fill` can stay like this for now.

Reply:
Turns out `where` lowering in inductor is fine, but inductor doesn't support `prims.to_dtype` on line 4765. Replacing it with `prims.convert_element_type` fixes test error even without `device_put` and makes generated code much nicer (with this decomp it falls back to prims.to_dtype which we definitely want to avoid, and generates a cpu kernel for a 0d cpu tensor, which is also very wasteful). 
@mruberry why do we have both `convert_element_type` and `to_dtype` and what's the difference between them? Should we remove `to_dtype`, its meta even explicitly says it's incorrect.

Reply:
Yes, removing `to_dtype` makes sense. It was a temporary addition to explore stride propagation

Reply:
Right, `_maybe_convert_to_dtype` is more appropriate, as conversion is not always needed.